processor/src/main/java/org/mapstruct/ap/processor/DefaultVersionInformation.java

```diff
@@ -0,0 +1,168 @@
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.jar.Manifest;
+
+import javax.annotation.processing.ProcessingEnvironment;
+
+import org.mapstruct.ap.version.VersionInformation;
+
+/**
+ * Provides information about the processor version and the processor context implementation version
+ *
+ * @author Andreas Gudian
+ */
+public class DefaultVersionInformation implements VersionInformation {
+    private static final String MAP_STRUCT_VERSION = initMapStructVersion();
+
+    private final String runtimeVersion;
+    private final String runtimeVendor;
+    private final String processingEnvironmentImplementationName;
+    private final String processingEnvironmentImplementationLibraryName;
+
+    DefaultVersionInformation(String runtimeVersion, String runtimeVendor,
+                              String processingEnvironmentImplementationName,
+                              String processingEnvironmentImplementationLibraryName) {
+        this.runtimeVersion = runtimeVersion;
+        this.runtimeVendor = runtimeVendor;
+        this.processingEnvironmentImplementationName = processingEnvironmentImplementationName;
+        this.processingEnvironmentImplementationLibraryName = processingEnvironmentImplementationLibraryName;
+    }
+
+    @Override
+    public String getRuntimeVersion() {
+        return this.runtimeVersion;
+    }
+
+    @Override
+    public String getRuntimeVendor() {
+        return this.runtimeVendor;
+    }
+
+    @Override
+    public String getMapStructVersion() {
+        return MAP_STRUCT_VERSION;
+    }
+
+    @Override
+    public String getProcessingEnvironmentImplementationName() {
+        return this.processingEnvironmentImplementationName;
+    }
+
+    @Override
+    public String getProcessingEnvironmentImplementationLibraryName() {
+        return this.processingEnvironmentImplementationLibraryName;
+    }
+
+    static DefaultVersionInformation fromProcessingEnvironment(ProcessingEnvironment processingEnv) {
+        String runtimeVersion = System.getProperty( "java.version" );
+        String runtimeVendor = System.getProperty( "java.vendor" );
+
+
+        String libraryFileName = getLibraryName( processingEnv );
+
+        return new DefaultVersionInformation(
+            runtimeVersion,
+            runtimeVendor,
+            processingEnv.getClass().getSimpleName(),
+            libraryFileName );
+    }
+
+    private static String getLibraryName(ProcessingEnvironment processingEnv) {
+        String classFileName = asClassFileName( processingEnv.getClass() );
+        URL resource = processingEnv.getClass().getClassLoader().getResource( classFileName );
+
+        if ( "jar".equals( resource.getProtocol() ) ) {
+            return extractJarFileName( resource.getFile() );
+        }
+        else if ( "jrt".equals( resource.getProtocol() ) ) {
+            // JDK 9 style, e.g. jrt:/jdk.compiler/com/sun/tools/javac/processing/JavacProcessingEnvironment.class
```

Comment:
Yes, we should. There's no JDK 9 available on Cloudbees, though, so I'll have to add an off-switch for that :wink:. I'll create a ticket for it.